default/python/AI/DiplomaticCorp.py

```diff
@@ -36,6 +36,42 @@ def handle_diplomatic_message(self, message):
         if message.recipient != fo.empireID():
             return
         aistate = get_aistate()
+        if message.type == fo.diplomaticMessageType.alliesProposal:
+            proposal_sender_player = fo.empirePlayerID(message.sender)
+            attitude = aistate.character.attitude_to_empire(message.sender, aistate.diplomatic_logs)
+            possible_acknowledgments = []
+            aggression = aistate.character.get_trait(Aggression)
+            if aggression.key == fo.aggression.beginner:
+                attitude = attitude + 50
+            elif aggression.key == fo.aggression.turtle:
+                attitude = attitude + 10
+            elif aggression.key == fo.aggression.cautious:
+                attitude = attitude + 5
+            if aggression.key <= fo.aggression.typical:
+                possible_acknowledgments = UserStringList("AI_ALLIANCE_PROPOSAL_ACKNOWLEDGEMENTS_MILD_LIST")
+                if attitude > 10:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_YES_MILD_LIST")
+                else:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_NO_MILD_LIST")
+            else:
+                possible_acknowledgments = UserStringList("AI_ALLIANCE_PROPOSAL_ACKNOWLEDGEMENTS_HARSH_LIST")
+                if attitude > 10:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_YES_HARSH_LIST")
+                else:
+                    possible_replies = UserStringList("AI_ALLIANCE_PROPOSAL_RESPONSES_NO_HARSH_LIST")
```

Comment:
The ```attitude``` calcs look out of place and seem error prone.

1) Current definition is that ```character.attitude_to_empire``` is limited to range [-10 .. 10]
2) The character already considers aggression trait when determining the empire's attitude.

Unless I am missing something, the current code is equivalent to the following (which I strongly prefer for clarity):
```
if aggression.key == fo.aggression.beginner:
    accept_proposal = True
elif aggression.key == fo.aggression.turtle:
    accept_proposal = attitude > 0
elif aggression.key == fo.aggression.cautious:
    accept_proposal = attitude > 5
else:  # aggression typical or greater
    accept_proposal = False
```

In particular, AIs with aggressiveness at least typical will never accept the alliance. That is not a gameplay problem considering the AI does not really handle allied status anyway, so preventing higher "difficulty" AI from doing something stupid is a good thing. However, that fact is not obvious from reading the proposed implementation.

Reply:
hopefully implemented...